tests/convergence/ex1c.cpp

```diff
@@ -0,0 +1,414 @@
+         umf_solver.Control[UMFPACK_ORDERING] = UMFPACK_ORDERING_METIS;
+         umf_solver.SetOperator(*A);
+         umf_solver.Mult(B, X);
+      }
+#endif
+      // Recover the solution as a finite element grid function.
+      a->RecoverFEMSolution(X, *b, x);
+
+      // Compute relative error
+      int order_quad = max(2, 2*order+1);
+      const IntegrationRule *irs[Geometry::NumGeom];
+      for (int i=0; i < Geometry::NumGeom; ++i)
+      {
+         irs[i] = &(IntRules.Get(i, order_quad));
+      }
+
+      double L2err_re  = x.real().ComputeL2Error(u_ex_re, irs);
+      double L2err_im  = x.imag().ComputeL2Error(u_ex_im, irs);
+
+      double H1err_re = x.real().ComputeH1Error(&u_ex_re, &u_grad_coeff_re,
+                                                &one, 1.0, 1);
+      double H1err_im = x.imag().ComputeH1Error(&u_ex_im, &u_grad_coeff_im,
+                                                &one, 1.0, 1);
+
+      double alpha0, beta0;
+      double alpha1, beta1;
+      if (l==0)
+      {
+         alpha0 = 0.0; beta0 = 0.0;
+         alpha1 = 0.0; beta1 = 0.0;
+      }
+      else
+      {
+         alpha0 = log(L2err_re0/L2err_re)/log(2.0);
+         beta0  = log(L2err_im0/L2err_im)/log(2.0);
+         alpha1 = log(H1err_re0/H1err_re)/log(2.0);
+         beta1  = log(H1err_im0/H1err_im)/log(2.0);
+      }
+
+      cout << setprecision(3);
+
+
+      cout << "real: || u_h - u ||_{H^1} = " << scientific
+           << H1err_re << ",  rate: " << fixed  << alpha1 << endl;
+      cout << "imag: || u_h - u ||_{H^1} = " << scientific
+           << H1err_im << ",  rate: " << fixed  << beta1 << endl;
+
+      cout << "real: || u_h - u ||_{L^2} = " << scientific
+           << L2err_re <<
+           ",  rate: " << fixed << alpha0 << endl;
+      cout << "imag: || u_h - u ||_{L^2} = " << scientific
+           << L2err_im <<
+           ",  rate: " << fixed << beta0 << endl;
+      cout << endl;
+
+      L2err_re0 = L2err_re;
+      L2err_im0 = L2err_im;
+
+      H1err_re0 = H1err_re;
+      H1err_im0 = H1err_im;
+
+      if (l == ref_levels) { break; }
+
+      mesh->UniformRefinement();
+      fespace->Update();
+      a->Update();
+      b->Update();
+      x.Update();
+      x_ex.Update();
+   }
+
+   // Send the solution by socket to a GLVis server.
+   if (visualization)
+   {
+      char vishost[] = "localhost";
+      int  visport   = 19916;
+      socketstream sol_sock_re(vishost, visport);
+      sol_sock_re.precision(8);
+      sol_sock_re << "solution\n" << *mesh << x.real() << "keys cm\n"
+                  << "window_title 'Real Part' " << flush;
+      socketstream sol_sock_im(vishost, visport);
+      sol_sock_im.precision(8);
+      sol_sock_im << "solution\n" << *mesh << x.imag() << "keys cm\n"
+                  << "window_title 'Imag Part' " << flush;
+   }
+
+   // 15. Free the used memory.
+   delete a;
+   delete b;
+   delete fespace;
+   if (order > 0) { delete fec; }
+   delete mesh;
+   return 0;
```

Comment:
Since this is a test we should return a non-zero value when the convergence rates aren't within some tolerance of the expected rates.

Reply:
Fixed